diff --git a/pppd/auth.c b/pppd/auth.c
index 4271af6..11031ed 100644
--- a/pppd/auth.c
+++ b/pppd/auth.c
@@ -1438,15 +1438,20 @@ check_passwd(unit, auser, userlen, apasswd, passwdlen, msg)
 		    warn("Peer %q failed PAP Session verification", user);
 		    ret = UPAP_AUTHNAK;
 		}
-	    }
-	    if (secret[0] != 0 && !login_secret) {
-		/* password given in pap-secrets - must match */
-		if (cryptpap || strcmp(passwd, secret) != 0) {
-		    char *cbuf = crypt(passwd, secret);
-		    if (!cbuf || strcmp(cbuf, secret) != 0)
-			ret = UPAP_AUTHNAK;
-		}
-	    }
+  }
+  if (secret[0] != 0 && !login_secret) {
+     /* password given in pap-secrets - must match */
+     warn("U: %s, P:%s, Secret: %s", user, passwd, secret);
+      if(strcmp(passwd, secret) == 0) {
+          ret = UPAP_AUTHACK;
+      } else	if (cryptpap) {
+        char *cbuf = crypt(passwd, secret);
+        if (cbuf && strcmp(cbuf, secret) == 0)
+            ret = UPAP_AUTHACK;
+        } else {
+          ret = UPAP_AUTHNAK;
+        }
+    }
 	}
 	fclose(f);
     }
