diff -Nuar a/net/bridge/br_input.c b/net/bridge/br_input.c
--- a/net/bridge/br_input.c	2017-03-10 10:56:08.665990775 +0000
+++ b/net/bridge/br_input.c	2017-03-10 11:11:56.000000000 +0000
@@ -219,9 +219,11 @@
 	struct net_bridge_port *p = br_port_get_rcu(skb->dev);
 	u16 vid = 0;
 
-	/* check if vlan is allowed, to avoid spoofing */
-	if (p->flags & BR_LEARNING && br_should_learn(p, skb, &vid))
-		br_fdb_update(p->br, p, eth_hdr(skb)->h_source, vid, false);
+	if (p->state != BR_STATE_DISABLED) {
+		/* check if vlan is allowed, to avoid spoofing */
+		if (p->flags & BR_LEARNING && br_should_learn(p, skb, &vid))
+			br_fdb_update(p->br, p, eth_hdr(skb)->h_source, vid, false);
+	}
 	return 0;	 /* process further */
 }
 
@@ -296,6 +298,27 @@
 
 forward:
 	switch (p->state) {
+	/*allow wpa eapol packet when disabled*/
+	case BR_STATE_DISABLED:
+		if (eth_hdr(skb)->h_proto == htons(ETH_P_PAE)) {
+			if (ether_addr_equal(p->br->dev->dev_addr, dest)) {
+				skb->pkt_type = PACKET_HOST;
+			}
+
+			if (NF_HOOK(NFPROTO_BRIDGE, NF_BR_PRE_ROUTING, 
+				dev_net(skb->dev), NULL, skb, skb->dev, NULL,
+				br_handle_local_finish)) {
+				break;
+			}
+
+			BR_INPUT_SKB_CB(skb)->brdev = p->br->dev;
+			br_pass_frame_up(skb);
+		}
+		else {
+			goto drop;
+		}
+		break;
+		
 	case BR_STATE_FORWARDING:
 		rhook = rcu_dereference(br_should_route_hook);
 		if (rhook) {
